spec/integration/outputs/index_spec.rb

```diff
@@ -0,0 +1,129 @@
+# encoding: utf-8
+require "logstash/devutils/rspec/spec_helper"
+require "logstash/outputs/elastic_app_search"
+require "logstash/codecs/plain"
+require "logstash/event"
+require "json"
+
+describe "indexing against running AppSearch", :integration => true do
+
+  let(:engine_name) {
+    (0...10).map { ('a'..'z').to_a[rand(26)] }.join
+  }
+
+  let(:config) {
+    {
+      "api_key" => ENV['APPSEARCH_PRIVATE_KEY'],
+      "engine" => engine_name,
+      "url" => "http://appsearch:3002"
+    }
+  }
+
+  subject { LogStash::Outputs::ElasticAppSearch.new(config) }
+
+  before(:each) do
+    create_engine(engine_name, "http://appsearch:3002", ENV['APPSEARCH_PRIVATE_KEY'])
+  end
+
+  private
+  def create_engine(engine_name, host, api_key)
+    url = host + "/api/as/v1/engines"
+    resp = Faraday.post(url, "{\"name\": \"#{engine_name}\"}",
+      "Content-Type" => "application/json",
+      "Authorization" => "Bearer " + api_key)
+    expect(resp.status).to eq(200)
+  end
+
+  describe "search and private keys are configured" do
+    let(:api_key_settings) {
+      {
+        :private => ENV['APPSEARCH_PRIVATE_KEY'],
+        :search => ENV['APPSEARCH_SEARCH_KEY']
+      }
+    }
+
+    it "setup apikeys" do
+      expect(api_key_settings[:private]).to start_with("private-")
+      expect(api_key_settings[:search]).to start_with("search-")
+    end
+  end
+
+  describe "indexing" do
+
+    before do
+      subject.register
+    end
+
+    describe "single event" do
+      let(:event) {
+        LogStash::Event.new("message" => "an event to index")
+      }
+
+      it "should be indexed" do
+        subject.multi_receive([event])
+
+        response = wait_for_response(1) do
+          execute_search_call(engine_name)
+        end
+        expect(response.status).to eq(200)
+        parsed_resp = JSON.parse(response.body)
+        expect(parsed_resp["meta"]["page"]["total_results"]).to eq(1)
+        results = parsed_resp["results"]
+        expect(results[0]["message"]["raw"]).to eq("an event to index")
+      end
+    end
+
+    private
+    def wait_for_response(expected_docs_count = 1)
```

Comment:
The method name `wait_for_response` doesn't clearly indicate to me that we are retrying a request until the response has certain properties, and uses of this method all re-parse the response body and re-validate the `total_results`.

This method is called twice, and in both cases we pass an identical block (`execute_search_call(engine_name)`). In many other places in Logstash, we use `Stud::try` for this effect:

~~~ ruby
  it 'is indexed' do
    results = Stud.try(5.times) do
      attempt_response = execute_search_call(engine_name)
      expect(attempt_response.status).to eq(200)
      parsed_resp = JSON.parse(response.body)
      expect(parsed_resp.dig("meta", "page", "total_results")).to be >= 1
      parsed_resp["results"]
    end
    expect(results.first.dig("message", "raw")).to eq "an event to index"
  end
~~~

Separately, I'm using [`Hash#dig`](https://apidock.com/ruby/Hash/dig) here, which is a safer way of digging into a deeply nested hash structure since an intermediate missing value won't result in a `NoMethodError` sending `[]` to an instance of `NilClass`.